cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -0,0 +1,285 @@
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <arrow/compute/api.h>
+#include <arrow/util/bit_block_counter.h>
+#include <arrow/util/bitmap_ops.h>
+
+#include "codegen_internal.h"
+
+namespace arrow {
+using internal::BitBlockCount;
+using internal::BitBlockCounter;
+
+namespace compute {
+
+namespace {
+
+// nulls will be promoted as follows
+// cond.val && (cond.data && left.val || ~cond.data && right.val)
+Status promote_nulls(KernelContext* ctx, const ArrayData& cond, const ArrayData& left,
+                     const ArrayData& right, ArrayData* output) {
+  if (!cond.MayHaveNulls() && !left.MayHaveNulls() && !right.MayHaveNulls()) {
+    return Status::OK();  // no nulls to handle
+  }
+  const int64_t len = cond.length;
+
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Buffer> out_validity, ctx->AllocateBitmap(len));
+  arrow::internal::InvertBitmap(out_validity->data(), 0, len,
+                                out_validity->mutable_data(), 0);
+  if (right.MayHaveNulls()) {
+    // out_validity = right.val && ~cond.data
+    arrow::internal::BitmapAndNot(right.buffers[0]->data(), right.offset,
+                                  cond.buffers[1]->data(), cond.offset, len, 0,
+                                  out_validity->mutable_data());
+  }
+
+  if (left.MayHaveNulls()) {
+    // tmp_buf = left.val && cond.data
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Buffer> temp_buf,
+                          arrow::internal::BitmapAnd(
+                              ctx->memory_pool(), left.buffers[0]->data(), left.offset,
+                              cond.buffers[1]->data(), cond.offset, len, 0));
+    // out_validity = cond.data && left.val || ~cond.data && right.val
+    arrow::internal::BitmapOr(out_validity->data(), 0, temp_buf->data(), 0, len, 0,
+                              out_validity->mutable_data());
+  }
+
+  if (cond.MayHaveNulls()) {
+    // out_validity &= cond.val
+    ::arrow::internal::BitmapAnd(out_validity->data(), 0, cond.buffers[0]->data(),
+                                 cond.offset, len, 0, out_validity->mutable_data());
+  }
+
+  output->buffers[0] = std::move(out_validity);
+  output->GetNullCount();  // update null count
+  return Status::OK();
+}
+
+template <typename Type, bool swap = false, typename Enable = void>
```

Comment:
What is `swap` doing?

Reply:
My idea is to reuse the impl for the cases like, `cond, left: Array, right: Scalar` and `cond, left: Sca;ar, right: Array`. In the second scenario, I can swap left and right and invert the cond without changing the loop mechanism.